intelmq/lib/bot.py

```diff
@@ -264,14 +287,38 @@ def start(self, starting: bool = True, error_on_pipeline: bool = True,
                             # retry forever, see https://github.com/certtools/intelmq/issues/1333
                             # https://lists.cert.at/pipermail/intelmq-users/2018-October/000085.html
                             pass
+                else:
+                    self.__message_counter["success"] += 1
+                    # no errors, check for run mode: scheduled
+                    if self.run_mode == 'scheduled':
+                        self.logger.info('Shutting down scheduled bot.')
+                        self.stop(exitcode=0)
 
-                # no errors, check for run mode: scheduled
-                elif self.run_mode == 'scheduled':
-                    self.logger.info('Shutting down scheduled bot.')
-                    self.stop(exitcode=0)
-
+            self.__stats()
             self.__handle_sighup()
 
+    def __stats(self, force=False):
+        """
+        Flush stats to redis
+
+        Only all self.__message_counter_delay (2 seconds), or with force=True
+        """
+
+        if not (force or datetime.now() - self.__message_counter["stats_timestamp"] > self.__message_counter_delay):
+            return
+
+        try:
+            for path, n in self.__message_counter["path"].items():  # current queue traffic
+                self.__stats_cache.set(".".join((self.__bot_id, "temporary", path)), n, ttl=2)
+                self.__message_counter["path_total"][path] += n
+                self.__message_counter["path"][path] = 0
+            for path, n in self.__message_counter["path_total"].items():  # total queue traffic
+                self.__stats_cache.set(".".join((self.__bot_id, "total", path)), n)
```

Comment:
Since you're using Cache.set without ttl, the redis command will be SETNX – in other words, the initial value is never replaced. We want this value to endure even it the bot is stopped, right?

Reply:
Oh, thanks for finding this. It should be set of course. But also for all the caches IMO.

Reply:
ac66a916db36db996a8042778ccfabe4448c9b86